pkg/sources/reconciler/awssnssource/subscription.go

```diff
@@ -156,30 +147,63 @@ func (r *Reconciler) ensureUnsubscribed(ctx context.Context) error {
 
 	topicARN := typedSrc.Spec.ARN.String()
 
-	subsARN, err := findSubscription(ctx, snsClient, topicARN, sm.Address.URL.String())
+	// We need to check the existence of the topic prior to attempting to
+	// list its subscriptions, because AWS returns the same NotFound error
+	// whether the topic or the subscription is missing, and we want to
+	// handle both cases differently.
+	topicExists, err := checkTopicExists(ctx, snsClient, topicARN)
```

Comment:
In fact, we should also do that inside `ensureSubscribe()`. Currently, we end up in a reconciliation loop if the topic doesn't exist.

Let me know if you're fine with me adding it in this PR, even though it's not directly related to the original issue.

Reply:
Do what you've got to do.